hoomd/ParticleGroup.cc

```diff
@@ -789,6 +792,100 @@ unsigned int ParticleGroup::intersectionSize(std::shared_ptr<ParticleGroup> othe
+                                  access_mode::readwrite);
+
+    ArrayHandle<Scalar3> h_inertia(m_pdata->getMomentsOfInertiaArray(),
+                                   access_location::host,
+                                   access_mode::read);
+
+    ArrayHandle<unsigned int> h_tag(m_pdata->getTags(),
+                                    access_location::host,
+                                    access_mode::read);
+
+    // Total the system's linear momentum
+    vec3<Scalar> tot_momentum(0, 0, 0);
+
+    // Loop over all particles in the group
+    for (unsigned int group_idx = 0; group_idx < group_size; group_idx++)
+        {
+        unsigned int j = this->getMemberIndex(group_idx);
+        unsigned int ptag = h_tag.data[j];
+
+        // Seed the RNG
+        hoomd::RandomGenerator rng(hoomd::RNGIdentifier::ParticleGroupThermalize, seed, ptag, timestep);
+
+        // Generate a random velocity
+        Scalar mass =  h_vel.data[j].w;
+        Scalar sigma = fast::sqrt(kT / mass);
+        hoomd::NormalDistribution<Scalar> normal(sigma);
+        h_vel.data[j].x = normal(rng);
+        h_vel.data[j].y = normal(rng);
+        if (D > 2)
+            h_vel.data[j].z = normal(rng);
+        else
+            h_vel.data[j].z = 0; // For 2D systems
+
+        tot_momentum += mass * vec3<Scalar>(h_vel.data[j]);
+
+        // Generate random angular momentum if the particle has rotational degrees of freedom.
+        vec3<Scalar> p_vec(0,0,0);
+        quat<Scalar> q(h_orientation.data[j]);
+        vec3<Scalar> I(h_inertia.data[j]);
+
+        if (I.x > 0)
+            p_vec.x = hoomd::NormalDistribution<Scalar>(fast::sqrt(kT * I.x))(rng);
+        if (I.y > 0)
+            p_vec.y = hoomd::NormalDistribution<Scalar>(fast::sqrt(kT * I.y))(rng);
+        if (I.z > 0)
+            p_vec.z = hoomd::NormalDistribution<Scalar>(fast::sqrt(kT * I.z))(rng);
+
+        // Store the angular momentum quaternion
+        quat<Scalar> p = Scalar(2.0) * q * p_vec;
+        h_angmom.data[j] = quat_to_scalar4(p);
+        }
+
+    // Remove the center of mass momentum
+
+    #ifdef ENABLE_MPI
+    // Reduce the total momentum from all MPI ranks
+    if (m_pdata->getDomainDecomposition())
+       {
+       MPI_Allreduce(MPI_IN_PLACE, &tot_momentum, 3, MPI_HOOMD_SCALAR,
+                     MPI_SUM, m_exec_conf->getMPICommunicator());
+       }
+    #endif
+
+    vec3<Scalar> com_momentum(tot_momentum /
+                              Scalar(this->getNumMembersGlobal()));
+
+    for (unsigned int group_idx = 0; group_idx < group_size; group_idx++)
+        {
+        unsigned int j = this->getMemberIndex(group_idx);
+        Scalar mass =  h_vel.data[j].w;
+        h_vel.data[j].x = h_vel.data[j].x - com_momentum.x / mass;
+        h_vel.data[j].y = h_vel.data[j].y - com_momentum.y / mass;
+        if (D > 2)
+            h_vel.data[j].z = h_vel.data[j].z - com_momentum.z / mass;
+        else
+            h_vel.data[j].z = 0; // For 2D systems
```

Comment:
Same ternary comment as above (either apply the change in both places or neither depending on your stylistic preference).